File: 572. Subtree of Another Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
  def isSubtree(self, root, subRoot):
    """
    :type root: TreeNode
    :type subRoot: TreeNode
    :rtype: bool
    """
    def dfs(current, subCurrent):
      if current == None:
        return False
      elif isIdenticalDFS(current, subCurrent):
        return True

      return dfs(current.left, subCurrent) or dfs(current.right, subCurrent)
    
    def isIdenticalDFS(current, subCurrent):
      if current == None or subCurrent == None:
        return current == None and subCurrent == None
      return current.val == subCurrent.val and isIdenticalDFS(current.left, subCurrent.left) and isIdenticalDFS(current.right, subCurrent.right)

    return dfs(root, subRoot)

File: 572. Subtree of Another Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  boolean dfs(TreeNode current, TreeNode subCurrent){
    if (current == null){
      return false;
    }
    else if (isIdenticalDFS(current, subCurrent)){
      return true;
    }

    return dfs(current.left, subCurrent) || dfs(current.right, subCurrent);
  }
  
  boolean isIdenticalDFS(TreeNode current, TreeNode subCurrent){
    if (current == null || subCurrent == null){
      return current == null && subCurrent == null;
    }
    return current.val == subCurrent.val && isIdenticalDFS(current.left, subCurrent.left) && isIdenticalDFS(current.right, subCurrent.right);
  }
  public boolean isSubtree(TreeNode root, TreeNode subRoot) {
    return dfs(root, subRoot);
  }
}